config.go

```diff
@@ -28,6 +28,23 @@ type Env struct {
 	Value *string `yaml:"value"`
 }
 
+type Options struct {
+	Scale       *ScaleOptions       `yaml:"scale,omitempty"`
+	Concurrency *ConcurrencyOptions `yaml:"concurrency,omitempty"`
+}
+
+type ScaleOptions struct {
+	Min    *int64  `yaml:"min,omitempty"`
+	Max    *int64  `yaml:"max,omitempty"`
+	Metric *string `yaml:"metric,omitempty"`
+}
+
+type ConcurrencyOptions struct {
```

Comment:
With the introduction of `metric` above, is this really still limited to `concurrency`? I'd guess only `Limit` is?

Reply:
How would you map those values into the config then so it is not bloated and easy to read? I am open to any suggestions :)

Reply:
maybe `target` and `utilization` should move into the scale options (that's what they are after all), and the concurrency limit should get a special place somehow (as it also is in Knative: It's inferred as an autoscaling setting (inferred into `target` actually) but also significant as a routing and traffic behavior setting)

Reply:
Actually, I think one design goal would be to allow to map the config names here directly to `kn` options. Unfortunately, we have chosen `--concurrency-target` and `--concurrency-utilization`, so its not so easy to move those around without confusing this UX. Of course we could do this hand-in-hand with a corresponding change on `kn`, but this would be a more tricky because of backwards compatibility issues ...

Reply:
> Actually, I think one design goal would be to allow to map the config names here directly to `kn` options. Unfortunately, we have chosen `--concurrency-target` and `--concurrency-utilization`, so its not so easy to move those around without confusing this UX.

```yaml
options:
  scaling:
    initial: 2
    min: 1
    max:  10
    window: 10s
  concurrency:
    target: 75
    utilization: 75
    limit: 100
```

Have we not already made enough of a leap with the YAML in this format to assume the user experience does not have a direct correlation to `kn` without at least a little interpretation? The function development experience is meant to be clear, concise, opinionated, and targeted towards developers new to Knative. Does a strict adherence to `kn` command semantics necessarily help this goal? I'm not sure. :shrug: 

I know that, as a Linux user I can never remember if I should use `adduser` or `useradd`, so I understand the UX concern! :smile: 

> Of course we could do this hand-in-hand with a corresponding change on kn, but this would be a more tricky because of backwards compatibility issues ...

What if we think of `func` (and any other `kn` plugin, I suppose) as testing ground for user experiences that __may__ make their way into `kn`, with proper deprecation timelines and `--experimental` flags? So that instead of needing to make a hand-in-hand corresponding change, there is a process for `kn` in adopting features/usability/etc from plugins.

Reply:
Ack, I don't think we need to inherit all design choices from `kn` (or even Knative!) if we can do better.

Reply:
@markusthoemmes so you'd like to have the config like this?
```
name: test
namespace: ""
runtime: go
...
options:
  scale:
    min: 0
    max: 10
    metric: rps
    target: 75
    utilization: 75
  concurrency:
    limit: 100
```

Reply:
> Ack, I don't think we need to inherit all design choices from `kn` (or even Knative!) if we can do better.

We can do better indeed!  We have the benefits of both hindsight and a higher level of abstraction, thanks in no small part to kn and Knative in general.  I most certainly like the clarity and simplicity of these options as they are shaping up here.

Reply:
Ok, fine with it and let's fix in on `kn` and also support other types of scaling then only concurrency (rps, cpu)

Just to be sure about this particular change: `concurrency.limit` is really only about the concurrency while `scale.target` is about any of concurrency, rps or cpu depending of which scaling type has been selected with the annotation `autoscaling.knative.dev/metric`. How would one specify "hard limits" then for RPS or CPU based scaling ? (if we stay with the analogy of soft and hard limits for target/limit). Is it also via `...spec.containerConcurrency` (then that field would be misnamed) or is this just not be possible ? If it is the former case, then I would put the `concurrency.limit` also to the scale group, as it feels unnatural to have target and limit in different sections.

Reply:
> Ok, fine with it and let's fix in on kn and also support other types of scaling then only concurrency (rps, cpu)

Introducing the scaling metric (`--scale-metric`) is then also a good hook for changing the `kn` option name to `--scale-target`.

I still believe having user-faced wording aligned across abstractions is a good thing.

Reply:
There are no hard limits for RPS. CPU limits would be set via the container's resource block.

I would not conflate those though. The concurrency limit is not just impacting scale, but it's also impacting traffic behavior. Likewise, CPU-limits are not only impacting scale but also the behavior of the container. Target on the other hand, is strictly a scaling concern.

Reply:
I have moved `target` and `utilization` to `scale` section in https://github.com/boson-project/func/pull/374/commits/7a8b7ef6bcf4a13c1f6858190b7db8fd8dedafa6

Reply:
Ok, I understand. Then I wonder why not then grouping all limits together ? Like something as:

```yaml
resources:
  limits:
    cpu: ....
    memory: ....
    concurrency: ....
  requests:
    cpu: ...
    memory: ...
```

@zroubalik How do you specify the resource limits currently for the container ?

Reply:
Resource requests/limits are not being exposed at the moment. But my idea was to add them once we agree on the scaling options and my intention was to add them to the options exactly the way as you proposed. The concurrency added to limits seems nice! Great idea.

Reply:
Nice idea indeed! I like!

Reply:
I have removed `concurrency.limits` from this PR and will handle that in another one together with resource requests/limits.